util/comm-mec_lpc.c

```diff
@@ -0,0 +1,224 @@
+{
+	uint8_t sum = 0;
+	for (int i = 0; i < size; ++i) {
+		sum += data[i];
+	}
+	return sum;
+};
+
+static int ec_command_lpc_mec_3(int command, int version, const void *outdata,
+				int outsize, void *indata, int insize)
+{
+	uint8_t csum = 0;
+	int i;
+
+	union {
+		struct ec_host_request rq;
+		char data[EC_LPC_HOST_PACKET_SIZE];
+	} u;
+
+	union {
+		struct ec_host_response rs;
+		char data[EC_LPC_HOST_PACKET_SIZE];
+	} r;
+
+	/* Fail if output size is too big */
+	if (outsize + sizeof(u.rq) > EC_LPC_HOST_PACKET_SIZE)
+		return -EC_RES_REQUEST_TRUNCATED;
+
+	/* Fill in request packet */
+	/* TODO(crosbug.com/p/23825): This should be common to all protocols */
+	u.rq.struct_version = EC_HOST_REQUEST_VERSION;
+	u.rq.checksum = 0;
+	u.rq.command = command;
+	u.rq.command_version = version;
+	u.rq.reserved = 0;
+	u.rq.data_len = outsize;
+
+	memcpy(&u.data[sizeof(u.rq)], outdata, outsize);
+	csum = ec_checksum_buffer(u.data, outsize + sizeof(u.rq));
+	u.rq.checksum = (uint8_t)(-csum);
+
+	if (wait_for_ec(EC_LPC_ADDR_HOST_CMD, 1000000)) {
+		fprintf(stderr, "Timeout waiting for EC response\n");
+		return -EC_RES_ERROR;
+	}
+
+	ec_mec_xfer(EC_MEC_WRITE, 0, u.data, outsize + sizeof(u.rq));
+
+	/* Start the command */
+	outb(EC_COMMAND_PROTOCOL_3, EC_LPC_ADDR_HOST_CMD);
+
+	if (wait_for_ec(EC_LPC_ADDR_HOST_CMD, 1000000)) {
+		fprintf(stderr, "Timeout waiting for EC response\n");
+		return -EC_RES_ERROR;
+	}
+
+	/* Check result */
+	i = inb(EC_LPC_ADDR_HOST_DATA);
+	if (i) {
+		fprintf(stderr, "EC returned error result code %d\n", i);
+		return -EECRESULT - i;
+	}
+
+	csum = 0;
+	ec_mec_xfer(EC_MEC_READ, 0, r.data, sizeof(r.rs));
+
+	if (r.rs.struct_version != EC_HOST_RESPONSE_VERSION) {
+		fprintf(stderr, "EC response version mismatch\n");
+		return -EC_RES_INVALID_RESPONSE;
+	}
+
+	if (r.rs.reserved) {
+		fprintf(stderr, "EC response reserved != 0\n");
+		return -EC_RES_INVALID_RESPONSE;
+	}
+
+	if (r.rs.data_len > insize) {
+		fprintf(stderr, "EC returned too much data\n");
+		return -EC_RES_RESPONSE_TOO_BIG;
+	}
+
+	if (r.rs.data_len > 0) {
+		ec_mec_xfer(EC_MEC_READ, 8, r.data + sizeof(r.rs), r.rs.data_len);
+		if (ec_checksum_buffer(r.data, sizeof(r.rs) + r.rs.data_len)) {
+			fprintf(stderr, "EC response has invalid checksum\n");
+			return -EC_RES_INVALID_CHECKSUM;
+		}
+
+		memcpy(indata, r.data + sizeof(r.rs), r.rs.data_len);
+	}
+	return r.rs.data_len;
+}
+
+static int ec_readmem_lpc_mec(int offset, int bytes, void *dest)
+{
+	int xfer_offset = MEC_EC_MEMMAP_START + offset;
+	int cnt = 0;
+	char* s = dest;
+
+	if (offset >= EC_MEMMAP_SIZE - bytes)
+		return -1;
+
+	if (bytes) {
+		ec_mec_xfer(EC_MEC_READ, xfer_offset, dest, bytes);
+		cnt = bytes;
+	} else {
+		/* Somewhat brute-force to set up a bunch of
+		 * individual transfers, but clearer than copying the xfer code
+		 * to add a stop condition.
+		 */
+		do {
+			ec_mec_xfer(EC_MEC_READ, xfer_offset++, s++, 1);
+			cnt++;
+		} while(*s != '\0' && xfer_offset < EC_MEMMAP_SIZE);
```

Comment:
this condition is incorrect; `xfer_offset` is already adjusted by `MEC_EC_MEMMAP_START` and this will always be fall of range

Reply:
(Fixed)